elf_reader.hpp

```diff
@@ -7,57 +7,85 @@
 enum class ReaderStatus : int {
     SUCCESS = 0,
     BAD_FILE = 1,
-    BAD_SEGMENT = 2
+    BAD_SEGMENT = 2,
+    BAD_SECTION = 3,
+    BAD_ALLOC = 4
 };
 
 class ElfReader {
     public:
-        ElfReader(std::string filename): reader(), file_loaded_(false) {
+        ElfReader(std::string filename): reader(), vmem_(nullptr), file_loaded_(false), max_vaddr_(0) {
             file_loaded_ = reader.load(filename);
         };
 
+        ~ElfReader() {
+            delete[] vmem_;
+        }
+
         ReaderStatus load_instructions(Hart& hart) {
             if (!file_loaded_) {
                 return ReaderStatus::BAD_FILE;
             }
 
-            char* data = nullptr;
-            uint64_t segment_start = 0, segment_end = 0;
             for (const auto& segment : reader.segments) {
-                if (segment->get_type() == ELFIO::PT_LOAD && (segment->get_flags() & ELFIO::PF_X)) {
-                    segment_start = segment->get_virtual_address();
-                    segment_end = segment_start + segment->get_memory_size();
+                if ((segment->get_type() == ELFIO::PT_LOAD) &&
+                    (segment->get_flags() & ELFIO::PF_X)) {
```

Comment:
does that mean we only load executable pages? if so, why?

Reply:
fixed